// vscode/vscode/src/vs/platform/agentHost/common/remoteAgentHostService.js

import { Event } from '../../../base/common/event.js';
import '../../../base/common/network.js';
import { TUNNEL_ADDRESS_PREFIX } from './tunnelAgentHost.js';

var RemoteAgentHostConnectionStatus;
(function(RemoteAgentHostConnectionStatus) {
    RemoteAgentHostConnectionStatus["Connected"] = "connected";
    RemoteAgentHostConnectionStatus["Connecting"] = "connecting";
    RemoteAgentHostConnectionStatus["Disconnected"] = "disconnected";
})(RemoteAgentHostConnectionStatus || (RemoteAgentHostConnectionStatus = {}));
const RemoteAgentHostsSettingId = "chat.remoteAgentHosts";
const RemoteAgentHostsEnabledSettingId = "chat.remoteAgentHostsEnabled";
var RemoteAgentHostEntryType;
(function(RemoteAgentHostEntryType) {
    RemoteAgentHostEntryType["WebSocket"] = "websocket";
    RemoteAgentHostEntryType["SSH"] = "ssh";
    RemoteAgentHostEntryType["Tunnel"] = "tunnel";
})(RemoteAgentHostEntryType || (RemoteAgentHostEntryType = {}));
function getEntryAddress(entry) {
    switch (entry.connection.type) {
    case RemoteAgentHostEntryType.WebSocket:
    case RemoteAgentHostEntryType.SSH:
        return entry.connection.address;
    case RemoteAgentHostEntryType.Tunnel:
        return `${TUNNEL_ADDRESS_PREFIX}${entry.connection.tunnelId}`;
    }
}
var RemoteAgentHostInputValidationError;
(function(RemoteAgentHostInputValidationError) {
    RemoteAgentHostInputValidationError["Empty"] = "empty";
    RemoteAgentHostInputValidationError["Invalid"] = "invalid";
})(
    RemoteAgentHostInputValidationError || (RemoteAgentHostInputValidationError = {})
);
class NullRemoteAgentHostService {
    constructor() {
        this.onDidChangeConnections = Event.None;
        this.connections = [];
        this.configuredEntries = [];
    }
    getConnection() {
        return undefined;
    }
    async addRemoteAgentHost() {
        throw ( new Error("Remote agent host connections are not supported in this environment."));
    }
    async removeRemoteAgentHost(_address) {}
    reconnect(_address) {}
    async addSSHConnection() {
        throw ( new Error("Remote agent host connections are not supported in this environment."));
    }
    getEntryByAddress() {
        return undefined;
    }
}
function rawEntryToEntry(raw) {
    if (raw.sshConfigHost || raw.sshHostName || raw.sshUser || raw.sshPort) {
        return {
            name: raw.name,
            connectionToken: raw.connectionToken,
            connection: {
                type: RemoteAgentHostEntryType.SSH,
                address: raw.address,
                sshConfigHost: raw.sshConfigHost,
                hostName: raw.sshHostName ?? raw.address,
                user: raw.sshUser,
                port: raw.sshPort
            }
        };
    }
    return {
        name: raw.name,
        connectionToken: raw.connectionToken,
        connection: {
            type: RemoteAgentHostEntryType.WebSocket,
            address: raw.address
        }
    };
}
function entryToRawEntry(entry) {
    switch (entry.connection.type) {
    case RemoteAgentHostEntryType.SSH:
        return {
            address: entry.connection.address,
            name: entry.name,
            connectionToken: entry.connectionToken,
            sshConfigHost: entry.connection.sshConfigHost,
            sshHostName: entry.connection.hostName,
            sshUser: entry.connection.user,
            sshPort: entry.connection.port
        };
    case RemoteAgentHostEntryType.WebSocket:
        return {
            address: entry.connection.address,
            name: entry.name,
            connectionToken: entry.connectionToken
        };
    case RemoteAgentHostEntryType.Tunnel:
        return undefined;
    }
}

export { NullRemoteAgentHostService, RemoteAgentHostConnectionStatus, RemoteAgentHostEntryType, RemoteAgentHostInputValidationError, RemoteAgentHostsEnabledSettingId, RemoteAgentHostsSettingId, entryToRawEntry, getEntryAddress, rawEntryToEntry };
