// @codingame/monaco-vscode-base-service-override/vscode/src/vs/workbench/contrib/inlineCompletions/browser/renameSymbolTrackerService.js

import { __decorate, __param } from '@codingame/monaco-vscode-api/external/tslib/tslib.es6';
import { Disposable } from '@codingame/monaco-vscode-api/vscode/vs/base/common/lifecycle';
import '@codingame/monaco-vscode-api/vscode/vs/base/common/observableInternal/index';
import { observableCodeEditor } from '@codingame/monaco-vscode-api/vscode/vs/editor/browser/observableCodeEditor';
import { ICodeEditorService } from '@codingame/monaco-vscode-api/vscode/vs/editor/browser/services/codeEditorService.service';
import '@codingame/monaco-vscode-api/vscode/vs/platform/instantiation/common/instantiation';
import { Range } from '@codingame/monaco-vscode-api/vscode/vs/editor/common/core/range';
import { StandardTokenType } from '@codingame/monaco-vscode-api/vscode/vs/editor/common/encodedTokenAttributes';
import { IModelService } from '@codingame/monaco-vscode-api/vscode/vs/editor/common/services/model.service';
import '@codingame/monaco-vscode-api/vscode/vs/platform/instantiation/common/extensions';
import { observableValue } from '@codingame/monaco-vscode-api/vscode/vs/base/common/observableInternal/observables/observableValue';
import { derived } from '@codingame/monaco-vscode-api/vscode/vs/base/common/observableInternal/observables/derived';
import { autorun } from '@codingame/monaco-vscode-api/vscode/vs/base/common/observableInternal/reactions/autorun';

function isUserEdit(event) {
    if (event.isUndoing || event.isRedoing || event.isFlush) {
        return false;
    }
    for (const source of event.detailedReasons) {
        if (!isUserEditSource(source)) {
            return false;
        }
    }
    return event.detailedReasons.length > 0;
}
const userEditKinds = ( new Set([
    "type",
    "paste",
    "cut",
    "executeCommands",
    "executeCommand",
    "compositionType",
    "compositionEnd"
]));
function isUserEditSource(source) {
    const metadata = source.metadata;
    if (metadata.source !== "cursor") {
        return false;
    }
    const kind = metadata.kind;
    return ( userEditKinds.has(kind));
}
class ModelSymbolRenameTracker extends Disposable {
    constructor(_model) {
        super();
        this._model = _model;
        this._trackedWord = observableValue(this, undefined);
        this.trackedWord = this._trackedWord;
        this._capturedWord = undefined;
        this._lastWordBeforeEdit = undefined;
        this._pendingContentChange = false;
        this._lastCursorPosition = undefined;
        this._register(this._model.onDidChangeContent(e => {
            if (!isUserEdit(e)) {
                const position = this._lastCursorPosition;
                this.reset();
                if (position !== undefined) {
                    this.updateCursorPosition(position);
                }
                return;
            }
            this._pendingContentChange = true;
        }));
    }
    updateCursorPosition(position) {
        this._lastCursorPosition = position;
        const wordAtPosition = this._model.getWordAtPosition(position);
        if (!wordAtPosition) {
            this._lastWordBeforeEdit = undefined;
            this._pendingContentChange = false;
            return;
        }
        if (this._isPositionInComment(position)) {
            this._lastWordBeforeEdit = undefined;
            this._pendingContentChange = false;
            return;
        }
        const currentWord = {
            word: wordAtPosition.word,
            range: ( new Range(
                position.lineNumber,
                wordAtPosition.startColumn,
                position.lineNumber,
                wordAtPosition.endColumn
            )),
            position
        };
        const contentChanged = this._pendingContentChange;
        this._pendingContentChange = false;
        if (!contentChanged) {
            this._lastWordBeforeEdit = currentWord;
            return;
        }
        if (!this._capturedWord) {
            const originalWord = this._lastWordBeforeEdit ?? currentWord;
            this._capturedWord = {
                ...originalWord
            };
            this._trackedWord.set({
                model: this._model,
                originalWord: originalWord.word,
                originalPosition: originalWord.position,
                originalRange: originalWord.range,
                currentWord: currentWord.word,
                currentRange: currentWord.range
            }, undefined);
            this._lastWordBeforeEdit = currentWord;
            return;
        }
        const capturedWord = this._capturedWord;
        const isOnSameWord = this._rangesOverlap(capturedWord.range, currentWord.range) || this._isAdjacent(capturedWord.range, currentWord.range);
        if (isOnSameWord) {
            this._trackedWord.set({
                model: this._model,
                originalWord: capturedWord.word,
                originalPosition: capturedWord.position,
                originalRange: capturedWord.range,
                currentWord: currentWord.word,
                currentRange: currentWord.range
            }, undefined);
        } else {
            const originalWord = this._lastWordBeforeEdit ?? currentWord;
            this._capturedWord = {
                ...originalWord
            };
            this._trackedWord.set({
                model: this._model,
                originalWord: originalWord.word,
                originalPosition: originalWord.position,
                originalRange: originalWord.range,
                currentWord: currentWord.word,
                currentRange: currentWord.range
            }, undefined);
        }
        this._lastWordBeforeEdit = currentWord;
    }
    reset() {
        this._trackedWord.set(undefined, undefined);
        this._capturedWord = undefined;
        this._lastWordBeforeEdit = undefined;
        this._pendingContentChange = false;
        this._lastCursorPosition = undefined;
    }
    _isPositionInComment(position) {
        this._model.tokenization.tokenizeIfCheap(position.lineNumber);
        const tokens = this._model.tokenization.getLineTokens(position.lineNumber);
        const tokenIndex = tokens.findTokenIndexAtOffset(position.column - 1);
        const tokenType = tokens.getStandardTokenType(tokenIndex);
        return tokenType === StandardTokenType.Comment;
    }
    _rangesOverlap(a, b) {
        if (a.startLineNumber !== b.startLineNumber) {
            return false;
        }
        return !(a.endColumn < b.startColumn || b.endColumn < a.startColumn);
    }
    _isAdjacent(a, b) {
        if (a.startLineNumber !== b.startLineNumber) {
            return false;
        }
        return a.endColumn === b.startColumn || b.endColumn === a.startColumn;
    }
}
let RenameSymbolTrackerService = class RenameSymbolTrackerService extends Disposable {
    constructor(_codeEditorService, _modelService) {
        super();
        this._codeEditorService = _codeEditorService;
        this._modelService = _modelService;
        this._modelTrackers = ( new Map());
        this._editorFocusTrackingDisposables = ( new Map());
        this._focusedModelTracker = observableValue(this, undefined);
        this.trackedWord = derived(this, reader => {
            const tracker = this._focusedModelTracker.read(reader);
            return tracker?.trackedWord.read(reader);
        });
        for (const editor of this._codeEditorService.listCodeEditors()) {
            this._setupEditorTracking(editor);
        }
        this._register(this._codeEditorService.onCodeEditorAdd(editor => {
            this._setupEditorTracking(editor);
        }));
        this._register(this._codeEditorService.onCodeEditorRemove(editor => {
            const focusDisposable = this._editorFocusTrackingDisposables.get(editor);
            if (focusDisposable) {
                focusDisposable.dispose();
                this._editorFocusTrackingDisposables.delete(editor);
            }
        }));
        this._register(this._modelService.onModelRemoved(model => {
            const tracker = this._modelTrackers.get(model);
            if (tracker) {
                tracker.dispose();
                this._modelTrackers.delete(model);
            }
        }));
    }
    _setupEditorTracking(editor) {
        if (editor.isSimpleWidget) {
            return;
        }
        if (!( this._editorFocusTrackingDisposables.has(editor))) {
            const obsEditor = observableCodeEditor(editor);
            const focusDisposable = autorun(reader => {
                const isFocused = obsEditor.isFocused.read(reader);
                const model = obsEditor.model.read(reader);
                const cursorPosition = obsEditor.cursorPosition.read(reader);
                if (!isFocused || !model) {
                    return;
                }
                let tracker = this._modelTrackers.get(model);
                if (!tracker) {
                    tracker = ( new ModelSymbolRenameTracker(model));
                    this._modelTrackers.set(model, tracker);
                }
                if (this._focusedModelTracker.read(undefined) !== tracker) {
                    this._focusedModelTracker.set(tracker, undefined);
                }
                if (cursorPosition) {
                    tracker.updateCursorPosition(cursorPosition);
                }
            });
            this._editorFocusTrackingDisposables.set(editor, focusDisposable);
        }
    }
    dispose() {
        for (const tracker of ( this._modelTrackers.values())) {
            tracker.dispose();
        }
        this._modelTrackers.clear();
        for (const disposable of ( this._editorFocusTrackingDisposables.values())) {
            disposable.dispose();
        }
        this._editorFocusTrackingDisposables.clear();
        super.dispose();
    }
};
RenameSymbolTrackerService = ( __decorate([( __param(0, ICodeEditorService)), ( __param(1, IModelService))], RenameSymbolTrackerService));

export { RenameSymbolTrackerService };
