// @codingame/monaco-vscode-api/vscode/src/vs/editor/contrib/codeAction/browser/codeActionMenu.js

import '../../../../base/browser/ui/codicons/codiconStyles.js';
import { Codicon } from '../../../../base/common/codicons.js';
import { CodeActionKind } from '../common/types.js';
import '../../symbolIcons/browser/symbolIcons.js';
import { localize } from '../../../../nls.js';
import { ActionListItemKind } from '../../../../platform/actionWidget/browser/actionList.js';
import { HierarchicalKind } from '../../../../base/common/hierarchicalKind.js';

const uncategorizedCodeActionGroup = ( Object.freeze({
    kind: HierarchicalKind.Empty,
    title: ( localize(993, "More Actions..."))
}));
const codeActionGroups = ( Object.freeze([{
    kind: CodeActionKind.QuickFix,
    title: ( localize(994, "Quick Fix"))
}, {
    kind: CodeActionKind.RefactorExtract,
    title: ( localize(995, "Extract")),
    icon: Codicon.wrench
}, {
    kind: CodeActionKind.RefactorInline,
    title: ( localize(996, "Inline")),
    icon: Codicon.wrench
}, {
    kind: CodeActionKind.RefactorRewrite,
    title: ( localize(997, "Rewrite")),
    icon: Codicon.wrench
}, {
    kind: CodeActionKind.RefactorMove,
    title: ( localize(998, "Move")),
    icon: Codicon.wrench
}, {
    kind: CodeActionKind.SurroundWith,
    title: ( localize(999, "Surround With")),
    icon: Codicon.surroundWith
}, {
    kind: CodeActionKind.Source,
    title: ( localize(1000, "Source Action")),
    icon: Codicon.symbolFile
}, uncategorizedCodeActionGroup]));
function toMenuItems(inputCodeActions, showHeaders, keybindingResolver) {
    if (!showHeaders) {
        return ( inputCodeActions.map(action => {
            return {
                kind: ActionListItemKind.Action,
                item: action,
                group: uncategorizedCodeActionGroup,
                disabled: !!action.action.disabled,
                label: action.action.disabled || action.action.title,
                canPreview: !!action.action.edit?.edits.length
            };
        }));
    }
    const menuEntries = ( codeActionGroups.map(group => ({
        group,
        actions: []
    })));
    for (const action of inputCodeActions) {
        const kind = action.action.kind ? ( new HierarchicalKind(action.action.kind)) : HierarchicalKind.None;
        for (const menuEntry of menuEntries) {
            if (menuEntry.group.kind.contains(kind)) {
                menuEntry.actions.push(action);
                break;
            }
        }
    }
    const allMenuItems = [];
    for (const menuEntry of menuEntries) {
        if (menuEntry.actions.length) {
            allMenuItems.push({
                kind: ActionListItemKind.Header,
                group: menuEntry.group
            });
            for (const action of menuEntry.actions) {
                const group = menuEntry.group;
                allMenuItems.push({
                    kind: ActionListItemKind.Action,
                    item: action,
                    group: action.action.isAI ? {
                        title: group.title,
                        kind: group.kind,
                        icon: Codicon.sparkle
                    } : group,
                    label: action.action.title,
                    disabled: !!action.action.disabled,
                    keybinding: keybindingResolver(action.action)
                });
            }
        }
    }
    return allMenuItems;
}

export { toMenuItems };
