// vscode/vscode/src/vs/workbench/contrib/codeActions/browser/codeActionsContribution.js

import { __decorate, __param } from '../../../../../../../external/tslib/tslib.es6.js';
import { Emitter, Event } from '../../../../base/common/event.js';
import { HierarchicalKind } from '../../../../base/common/hierarchicalKind.js';
import { Disposable } from '../../../../base/common/lifecycle.js';
import { editorConfigurationBaseNode } from '../../../../editor/common/config/editorConfigurationSchema.js';
import { ILanguageFeaturesService } from '../../../../editor/common/services/languageFeatures.service.js';
import { codeActionCommandId, refactorCommandId, sourceActionCommandId } from '../../../../editor/contrib/codeAction/browser/codeAction.js';
import { CodeActionKind } from '../../../../editor/contrib/codeAction/common/types.js';
import { localize } from '../../../../nls.js';
import { ConfigurationScope, Extensions } from '../../../../platform/configuration/common/configurationRegistry.js';
import { IKeybindingService } from '../../../../platform/keybinding/common/keybinding.service.js';
import { Registry } from '../../../../platform/registry/common/platform.js';

const createCodeActionsAutoSave = description => {
    return {
        type: "string",
        enum: ["always", "explicit", "never", true, false],
        enumDescriptions: [( localize(
            7833,
            "Triggers Code Actions on explicit saves and auto saves triggered by window or focus changes."
        )), ( localize(7834, "Triggers Code Actions only when explicitly saved")), ( localize(7835, "Never triggers Code Actions on save")), ( localize(
            7836,
            "Triggers Code Actions only when explicitly saved. This value will be deprecated in favor of \"explicit\"."
        )), ( localize(
            7837,
            "Never triggers Code Actions on save. This value will be deprecated in favor of \"never\"."
        ))],
        default: "explicit",
        description: description
    };
};
const createNotebookCodeActionsAutoSave = description => {
    return {
        type: ["string", "boolean"],
        enum: ["explicit", "never", true, false],
        enumDescriptions: [( localize(7838, "Triggers Code Actions only when explicitly saved.")), ( localize(7839, "Never triggers Code Actions on save.")), ( localize(
            7840,
            "Triggers Code Actions only when explicitly saved. This value will be deprecated in favor of \"explicit\"."
        )), ( localize(
            7841,
            "Triggers Code Actions only when explicitly saved. This value will be deprecated in favor of \"never\"."
        ))],
        default: "explicit",
        description: description
    };
};
const codeActionsOnSaveSchema = {
    oneOf: [{
        type: "object",
        additionalProperties: {
            type: "string"
        }
    }, {
        type: "array",
        items: {
            type: "string"
        }
    }],
    markdownDescription: ( localize(
        7842,
        "Run Code Actions for the editor on save. Code Actions must be specified and the editor must not be shutting down. When {0} is set to `afterDelay`, Code Actions will only be run when the file is saved explicitly. Example: `\"source.organizeImports\": \"explicit\" `",
        "`#files.autoSave#`"
    )),
    type: ["object", "array"],
    additionalProperties: {
        type: "string",
        enum: ["always", "explicit", "never", true, false]
    },
    default: {},
    scope: ConfigurationScope.LANGUAGE_OVERRIDABLE
};
const editorConfiguration = ( Object.freeze({
    ...editorConfigurationBaseNode,
    properties: {
        "editor.codeActionsOnSave": codeActionsOnSaveSchema
    }
}));
const notebookCodeActionsOnSaveSchema = {
    oneOf: [{
        type: "object",
        additionalProperties: {
            type: "string"
        }
    }, {
        type: "array",
        items: {
            type: "string"
        }
    }],
    markdownDescription: ( localize(
        7843,
        "Run a series of Code Actions for a notebook on save. Code Actions must be specified and the editor must not be shutting down. When {0} is set to `afterDelay`, Code Actions will only be run when the file is saved explicitly. Example: `\"notebook.source.organizeImports\": \"explicit\"`",
        "`#files.autoSave#`"
    )),
    type: "object",
    additionalProperties: {
        type: ["string", "boolean"],
        enum: ["explicit", "never", true, false]
    },
    default: {}
};
const notebookEditorConfiguration = ( Object.freeze({
    ...editorConfigurationBaseNode,
    properties: {
        "notebook.codeActionsOnSave": notebookCodeActionsOnSaveSchema
    }
}));
let CodeActionsContribution = class CodeActionsContribution extends Disposable {
    constructor(keybindingService, languageFeatures) {
        super();
        this.languageFeatures = languageFeatures;
        this._onDidChangeSchemaContributions = this._register(( new Emitter()));
        this._allProvidedCodeActionKinds = [];
        this._register(Event.runAndSubscribe(
            Event.debounce(languageFeatures.codeActionProvider.onDidChange, () => {}, 1000),
            () => {
                this._allProvidedCodeActionKinds = this.getAllProvidedCodeActionKinds();
                this.updateConfigurationSchema(this._allProvidedCodeActionKinds);
                this._onDidChangeSchemaContributions.fire();
            }
        ));
        this._register(keybindingService.registerSchemaContribution({
            getSchemaAdditions: () => this.getKeybindingSchemaAdditions(),
            onDidChange: this._onDidChangeSchemaContributions.event
        }));
    }
    getAllProvidedCodeActionKinds() {
        const out = ( new Map());
        for (const provider of this.languageFeatures.codeActionProvider.allNoModel()) {
            for (const kind of provider.providedCodeActionKinds ?? []) {
                out.set(kind, ( new HierarchicalKind(kind)));
            }
        }
        return Array.from(( out.values()));
    }
    updateConfigurationSchema(allProvidedKinds) {
        const properties = {
            ...codeActionsOnSaveSchema.properties
        };
        const notebookProperties = {
            ...notebookCodeActionsOnSaveSchema.properties
        };
        for (const codeActionKind of allProvidedKinds) {
            if (CodeActionKind.Source.contains(codeActionKind) && !properties[codeActionKind.value]) {
                properties[codeActionKind.value] = createCodeActionsAutoSave(( localize(
                    7844,
                    "Controls whether '{0}' actions should be run on file save.",
                    codeActionKind.value
                )));
                notebookProperties[codeActionKind.value] = createNotebookCodeActionsAutoSave(( localize(
                    7844,
                    "Controls whether '{0}' actions should be run on file save.",
                    codeActionKind.value
                )));
            }
        }
        codeActionsOnSaveSchema.properties = properties;
        notebookCodeActionsOnSaveSchema.properties = notebookProperties;
        ( Registry.as(Extensions.Configuration)).notifyConfigurationSchemaUpdated(editorConfiguration);
    }
    getKeybindingSchemaAdditions() {
        const conditionalSchema = (command, kinds) => {
            return {
                if: {
                    required: ["command"],
                    properties: {
                        "command": {
                            const: command
                        }
                    }
                },
                then: {
                    properties: {
                        "args": {
                            required: ["kind"],
                            properties: {
                                "kind": {
                                    anyOf: [{
                                        enum: Array.from(kinds)
                                    }, {
                                        type: "string"
                                    }]
                                }
                            }
                        }
                    }
                }
            };
        };
        const filterProvidedKinds = ofKind => {
            const out = ( new Set());
            for (const providedKind of this._allProvidedCodeActionKinds) {
                if (ofKind.contains(providedKind)) {
                    out.add(providedKind.value);
                }
            }
            return Array.from(out);
        };
        return [
            conditionalSchema(codeActionCommandId, filterProvidedKinds(HierarchicalKind.Empty)),
            conditionalSchema(refactorCommandId, filterProvidedKinds(CodeActionKind.Refactor)),
            conditionalSchema(sourceActionCommandId, filterProvidedKinds(CodeActionKind.Source))
        ];
    }
};
CodeActionsContribution = ( __decorate([( __param(0, IKeybindingService)), ( __param(1, ILanguageFeaturesService))], CodeActionsContribution));

export { CodeActionsContribution, editorConfiguration, notebookEditorConfiguration };
