// vscode/vscode/src/vs/platform/theme/common/colors/menuColors.js

import { localize } from '../../../../nls.js';
import { registerColor, transparent } from '../colorUtils.js';
import { contrastBorder, activeContrastBorder, foreground } from './baseColors.js';
import { selectForeground, selectBackground } from './inputColors.js';
import { listActiveSelectionForeground, listActiveSelectionBackground } from './listColors.js';

const menuBorder = registerColor("menu.border", {
    dark: null,
    light: null,
    hcDark: contrastBorder,
    hcLight: contrastBorder
}, ( localize(2370, "Border color of menus.")));
const menuForeground = registerColor("menu.foreground", selectForeground, ( localize(2371, "Foreground color of menu items.")));
const menuBackground = registerColor("menu.background", selectBackground, ( localize(2372, "Background color of menu items.")));
registerColor("menu.selectionForeground", listActiveSelectionForeground, ( localize(2373, "Foreground color of the selected menu item in menus.")));
registerColor("menu.selectionBackground", listActiveSelectionBackground, ( localize(2374, "Background color of the selected menu item in menus.")));
const menuSelectionBorder = registerColor("menu.selectionBorder", {
    dark: null,
    light: null,
    hcDark: activeContrastBorder,
    hcLight: activeContrastBorder
}, ( localize(2375, "Border color of the selected menu item in menus.")));
const menuSeparatorBackground = registerColor("menu.separatorBackground", {
    dark: ( transparent(foreground, 0.2)),
    light: ( transparent(foreground, 0.2)),
    hcDark: contrastBorder,
    hcLight: contrastBorder
}, ( localize(2376, "Color of a separator menu item in menus.")));

export { menuBackground, menuBorder, menuForeground, menuSelectionBorder, menuSeparatorBackground };
