// vscode/vscode/src/vs/workbench/contrib/speech/common/speechService.js

import { localize } from '../../../../nls.js';
import { RawContextKey } from '../../../../platform/contextkey/common/contextkey.js';
import { language } from '../../../../base/common/platform.js';

const HasSpeechProvider = ( new RawContextKey("hasSpeechProvider", false, {
    type: "boolean",
    description: ( localize(12991, "A speech provider is registered to the speech service."))
}));
const SpeechToTextInProgress = ( new RawContextKey("speechToTextInProgress", false, {
    type: "boolean",
    description: ( localize(12992, "A speech-to-text session is in progress."))
}));
const TextToSpeechInProgress = ( new RawContextKey("textToSpeechInProgress", false, {
    type: "boolean",
    description: ( localize(12993, "A text-to-speech session is in progress."))
}));
var SpeechToTextStatus;
(function(SpeechToTextStatus) {
    SpeechToTextStatus[SpeechToTextStatus["Started"] = 1] = "Started";
    SpeechToTextStatus[SpeechToTextStatus["Recognizing"] = 2] = "Recognizing";
    SpeechToTextStatus[SpeechToTextStatus["Recognized"] = 3] = "Recognized";
    SpeechToTextStatus[SpeechToTextStatus["Stopped"] = 4] = "Stopped";
    SpeechToTextStatus[SpeechToTextStatus["Error"] = 5] = "Error";
})(SpeechToTextStatus || (SpeechToTextStatus = {}));
var TextToSpeechStatus;
(function(TextToSpeechStatus) {
    TextToSpeechStatus[TextToSpeechStatus["Started"] = 1] = "Started";
    TextToSpeechStatus[TextToSpeechStatus["Stopped"] = 2] = "Stopped";
    TextToSpeechStatus[TextToSpeechStatus["Error"] = 3] = "Error";
})(TextToSpeechStatus || (TextToSpeechStatus = {}));
var KeywordRecognitionStatus;
(function(KeywordRecognitionStatus) {
    KeywordRecognitionStatus[KeywordRecognitionStatus["Recognized"] = 1] = "Recognized";
    KeywordRecognitionStatus[KeywordRecognitionStatus["Stopped"] = 2] = "Stopped";
    KeywordRecognitionStatus[KeywordRecognitionStatus["Canceled"] = 3] = "Canceled";
})(KeywordRecognitionStatus || (KeywordRecognitionStatus = {}));
var AccessibilityVoiceSettingId;
(function(AccessibilityVoiceSettingId) {
    AccessibilityVoiceSettingId["SpeechTimeout"] = "accessibility.voice.speechTimeout";
    AccessibilityVoiceSettingId["AutoSynthesize"] = "accessibility.voice.autoSynthesize";
    AccessibilityVoiceSettingId["SpeechLanguage"] = "accessibility.voice.speechLanguage";
    AccessibilityVoiceSettingId["IgnoreCodeBlocks"] = "accessibility.voice.ignoreCodeBlocks";
})(AccessibilityVoiceSettingId || (AccessibilityVoiceSettingId = {}));
const SPEECH_LANGUAGE_CONFIG = AccessibilityVoiceSettingId.SpeechLanguage;
const SPEECH_LANGUAGES = {
    ["da-DK"]: {
        name: ( localize(12994, "Danish (Denmark)"))
    },
    ["de-DE"]: {
        name: ( localize(12995, "German (Germany)"))
    },
    ["en-AU"]: {
        name: ( localize(12996, "English (Australia)"))
    },
    ["en-CA"]: {
        name: ( localize(12997, "English (Canada)"))
    },
    ["en-GB"]: {
        name: ( localize(12998, "English (United Kingdom)"))
    },
    ["en-IE"]: {
        name: ( localize(12999, "English (Ireland)"))
    },
    ["en-IN"]: {
        name: ( localize(13000, "English (India)"))
    },
    ["en-NZ"]: {
        name: ( localize(13001, "English (New Zealand)"))
    },
    ["en-US"]: {
        name: ( localize(13002, "English (United States)"))
    },
    ["es-ES"]: {
        name: ( localize(13003, "Spanish (Spain)"))
    },
    ["es-MX"]: {
        name: ( localize(13004, "Spanish (Mexico)"))
    },
    ["fr-CA"]: {
        name: ( localize(13005, "French (Canada)"))
    },
    ["fr-FR"]: {
        name: ( localize(13006, "French (France)"))
    },
    ["hi-IN"]: {
        name: ( localize(13007, "Hindi (India)"))
    },
    ["it-IT"]: {
        name: ( localize(13008, "Italian (Italy)"))
    },
    ["ja-JP"]: {
        name: ( localize(13009, "Japanese (Japan)"))
    },
    ["ko-KR"]: {
        name: ( localize(13010, "Korean (South Korea)"))
    },
    ["nl-NL"]: {
        name: ( localize(13011, "Dutch (Netherlands)"))
    },
    ["pt-PT"]: {
        name: ( localize(13012, "Portuguese (Portugal)"))
    },
    ["pt-BR"]: {
        name: ( localize(13013, "Portuguese (Brazil)"))
    },
    ["ru-RU"]: {
        name: ( localize(13014, "Russian (Russia)"))
    },
    ["sv-SE"]: {
        name: ( localize(13015, "Swedish (Sweden)"))
    },
    ["tr-TR"]: {
        name: ( localize(13016, "Turkish (Türkiye)"))
    },
    ["zh-CN"]: {
        name: ( localize(13017, "Chinese (Simplified, China)"))
    },
    ["zh-HK"]: {
        name: ( localize(13018, "Chinese (Traditional, Hong Kong)"))
    },
    ["zh-TW"]: {
        name: ( localize(13019, "Chinese (Traditional, Taiwan)"))
    }
};
function speechLanguageConfigToLanguage(config, lang = language) {
    if (typeof config === "string") {
        if (config === "auto") {
            if (lang !== "en") {
                const langParts = lang.split("-");
                return speechLanguageConfigToLanguage(`${langParts[0]}-${(langParts[1] ?? langParts[0]).toUpperCase()}`);
            }
        } else {
            if (SPEECH_LANGUAGES[config]) {
                return config;
            }
        }
    }
    return "en-US";
}

export { AccessibilityVoiceSettingId, HasSpeechProvider, KeywordRecognitionStatus, SPEECH_LANGUAGES, SPEECH_LANGUAGE_CONFIG, SpeechToTextInProgress, SpeechToTextStatus, TextToSpeechInProgress, TextToSpeechStatus, speechLanguageConfigToLanguage };
