// @codingame/monaco-vscode-quickaccess-service-override/vscode/src/vs/platform/quickinput/browser/helpQuickAccess.js

import { __decorate, __param } from '@codingame/monaco-vscode-api/external/tslib/tslib.es6';
import { localize } from '@codingame/monaco-vscode-api/vscode/vs/nls';
import { Registry } from '@codingame/monaco-vscode-api/vscode/vs/platform/registry/common/platform';
import { DisposableStore } from '@codingame/monaco-vscode-api/vscode/vs/base/common/lifecycle';
import { IContextKeyService } from '@codingame/monaco-vscode-api/vscode/vs/platform/contextkey/common/contextkey.service';
import { IKeybindingService } from '@codingame/monaco-vscode-api/vscode/vs/platform/keybinding/common/keybinding.service';
import { Extensions } from '@codingame/monaco-vscode-api/vscode/vs/platform/quickinput/common/quickAccess';
import { IQuickInputService } from '@codingame/monaco-vscode-api/vscode/vs/platform/quickinput/common/quickInput.service';

var HelpQuickAccessProvider_1;
let HelpQuickAccessProvider = class HelpQuickAccessProvider {
    static {
        HelpQuickAccessProvider_1 = this;
    }
    static {
        this.PREFIX = "?";
    }
    constructor(quickInputService, keybindingService, contextKeyService) {
        this.quickInputService = quickInputService;
        this.keybindingService = keybindingService;
        this.contextKeyService = contextKeyService;
        this.registry = ( Registry.as(Extensions.Quickaccess));
    }
    provide(picker) {
        const disposables = ( new DisposableStore());
        disposables.add(picker.onDidAccept(() => {
            const [item] = picker.selectedItems;
            if (item) {
                this.quickInputService.quickAccess.show(item.prefix, {
                    preserveValue: true
                });
            }
        }));
        disposables.add(picker.onDidChangeValue(value => {
            const providerDescriptor = this.registry.getQuickAccessProvider(
                value.substr(HelpQuickAccessProvider_1.PREFIX.length),
                this.contextKeyService
            );
            if (providerDescriptor?.prefix && providerDescriptor.prefix !== HelpQuickAccessProvider_1.PREFIX) {
                this.quickInputService.quickAccess.show(providerDescriptor.prefix, {
                    preserveValue: true
                });
            }
        }));
        picker.items = this.getQuickAccessProviders().filter(p => p.prefix !== HelpQuickAccessProvider_1.PREFIX);
        return disposables;
    }
    getQuickAccessProviders() {
        const providers = this.registry.getQuickAccessProviders(this.contextKeyService).sort((providerA, providerB) => providerA.prefix.localeCompare(providerB.prefix)).flatMap(provider => this.createPicks(provider));
        return providers;
    }
    createPicks(provider) {
        return (provider.helpEntries.map(helpEntry => {
            const prefix = helpEntry.prefix || provider.prefix;
            const label = prefix || "…" ;
            return {
                prefix,
                label,
                keybinding: helpEntry.commandId ? this.keybindingService.lookupKeybinding(helpEntry.commandId) : undefined,
                ariaLabel: ( localize(2062, "{0}, {1}", label, helpEntry.description)),
                description: helpEntry.description
            };
        }));
    }
};
HelpQuickAccessProvider = HelpQuickAccessProvider_1 = ( __decorate([( __param(0, IQuickInputService)), ( __param(1, IKeybindingService)), ( __param(2, IContextKeyService))], HelpQuickAccessProvider));

export { HelpQuickAccessProvider };
