// @codingame/monaco-vscode-base-service-override/vscode/src/vs/platform/request/common/requestIpc.js

import { bufferToStream } from '@codingame/monaco-vscode-api/vscode/vs/base/common/buffer';
import '@codingame/monaco-vscode-api/vscode/vs/base/common/cancellation';
import { Event } from '@codingame/monaco-vscode-api/vscode/vs/base/common/event';

class RequestChannelClient {
    constructor(channel) {
        this.channel = channel;
        this.onDidCompleteRequest = Event.None;
    }
    async request(options, token) {
        const [res, buffer] = await this.channel.call("request", [options], token);
        return {
            res,
            stream: bufferToStream(buffer)
        };
    }
    async resolveProxy(url) {
        return this.channel.call("resolveProxy", [url]);
    }
    async lookupAuthorization(authInfo) {
        return this.channel.call("lookupAuthorization", [authInfo]);
    }
    async lookupKerberosAuthorization(url) {
        return this.channel.call("lookupKerberosAuthorization", [url]);
    }
    async loadCertificates() {
        return this.channel.call("loadCertificates");
    }
}

export { RequestChannelClient };
