// @codingame/monaco-vscode-api/vscode/src/vs/platform/agentHost/browser/nullAgentHostService.js

import { Event } from '../../../base/common/event.js';
import '../../../base/common/observableInternal/index.js';
import { constObservable } from '../../../base/common/observableInternal/observables/constObservable.js';

const notSupported = () => {
    throw ( new Error("Local agent host is not supported in the browser."));
};
class NullAgentHostService {
    constructor() {
        this.clientId = "";
        this.onAgentHostExit = Event.None;
        this.onAgentHostStart = Event.None;
        this.onDidNotification = Event.None;
        this.onDidAction = Event.None;
        this.authenticationPending = constObservable(false);
    }
    setAuthenticationPending(_pending) {}
    get rootState() {
        return notSupported();
    }
    getSubscription(_kind, _resource) {
        return notSupported();
    }
    getSubscriptionUnmanaged(_kind, _resource) {
        return undefined;
    }
    dispatch(_action) {
        notSupported();
    }
    async restartAgentHost() {
        notSupported();
    }
    async authenticate(_params) {
        return notSupported();
    }
    async listSessions() {
        return [];
    }
    async createSession(_config) {
        return notSupported();
    }
    async resolveSessionConfig(_params) {
        return notSupported();
    }
    async sessionConfigCompletions(_params) {
        return notSupported();
    }
    async startWebSocketServer() {
        return notSupported();
    }
    async disposeSession(_session) {}
    async createTerminal(_params) {
        notSupported();
    }
    async disposeTerminal(_terminal) {}
    async resourceList(_uri) {
        return notSupported();
    }
    async resourceRead(_uri) {
        return notSupported();
    }
    async resourceWrite(_params) {
        return notSupported();
    }
    async resourceCopy(_params) {
        return notSupported();
    }
    async resourceDelete(_params) {
        return notSupported();
    }
    async resourceMove(_params) {
        return notSupported();
    }
}

export { NullAgentHostService };
