// @codingame/monaco-vscode-api/vscode/src/vs/workbench/contrib/terminal/common/terminalStrings.js

import { localize2, localize } from '../../../../nls.js';

const terminalStrings = {
    terminal: ( localize(13896, "Terminal")),
    new: ( localize(13897, "New Terminal")),
    currentSessionCategory: ( localize(13898, "current session")),
    previousSessionCategory: ( localize(13899, "previous session")),
    typeTask: ( localize(13900, "Task")),
    typeLocal: ( localize(13901, "Local")),
    actionCategory: ( localize2(13902, "Terminal")),
    focus: ( localize2(13903, "Focus Terminal")),
    kill: {
        ...( localize2(13904, "Kill Terminal")),
        short: ( localize(13905, "Kill"))
    },
    moveToEditor: ( localize2(13906, "Move Terminal into Editor Area")),
    moveIntoNewWindow: ( localize2(13907, "Move Terminal into New Window")),
    newInNewWindow: ( localize2(13908, "New Terminal Window")),
    moveToTerminalPanel: ( localize2(13909, "Move Terminal into Panel")),
    changeIcon: ( localize2(13910, "Change Icon...")),
    changeColor: ( localize2(13911, "Change Color...")),
    split: {
        ...( localize2(13912, "Split Terminal")),
        short: ( localize(13913, "Split"))
    },
    unsplit: ( localize2(13914, "Unsplit Terminal")),
    rename: ( localize2(13915, "Rename...")),
    toggleSizeToContentWidth: ( localize2(13916, "Toggle Size to Content Width")),
    focusHover: ( localize2(13917, "Focus Hover")),
    newWithCwd: ( localize2(13918, "Create New Terminal Starting in a Custom Working Directory")),
    renameWithArgs: ( localize2(13919, "Rename the Currently Active Terminal")),
    scrollToPreviousCommand: ( localize2(13920, "Scroll to Previous Command")),
    scrollToNextCommand: ( localize2(13921, "Scroll to Next Command"))};

export { terminalStrings };
