// vscode/vscode/src/vs/base/browser/domStylesheets.js

import { Disposable, toDisposable, DisposableStore } from '../common/lifecycle.js';
import '../common/observableInternal/index.js';
import { isFirefox } from './browser.js';
import { isShadowRoot, createElement, getWindows, sharedMutationObserver } from './dom.js';
import { mainWindow } from './window.js';
import { autorun } from '../common/observableInternal/reactions/autorun.js';

const globalStylesheets = ( new Map());
function isGlobalStylesheet(node) {
    return ( globalStylesheets.has(node));
}
class WrappedStyleElement extends Disposable {
    constructor() {
        super(...arguments);
        this._currentCssStyle = "";
        this._styleSheet = undefined;
    }
    setStyle(cssStyle) {
        if (cssStyle === this._currentCssStyle) {
            return;
        }
        this._currentCssStyle = cssStyle;
        if (!this._styleSheet) {
            this._styleSheet = createStyleSheet(mainWindow.document.head, s => s.textContent = cssStyle, this._store);
        } else {
            this._styleSheet.textContent = cssStyle;
        }
    }
    dispose() {
        super.dispose();
        this._styleSheet = undefined;
    }
}
let shadowRootContainer;
function setContainerElement(container) {
    const root = container.getRootNode();
    if (isShadowRoot(root)) {
        shadowRootContainer = root;
    }
}
function createStyleSheet(container = mainWindow.document.head, beforeAppend, disposableStore) {
    const style = createElement("style");
    style.type = "text/css";
    style.media = "screen";
    style.setAttribute("data-vscode", "true");
    beforeAppend?.(style);
    container.appendChild(style);
    if (disposableStore) {
        disposableStore.add(toDisposable(() => style.remove()));
    }
    if (container === mainWindow.document.head) {
        const globalStylesheetClones = ( new Set());
        globalStylesheets.set(style, globalStylesheetClones);
        if (disposableStore) {
            disposableStore.add(toDisposable(() => globalStylesheets.delete(style)));
        }
        for (const {
            window: targetWindow,
            disposables
        } of getWindows()) {
            if (targetWindow === mainWindow) {
                continue;
            }
            const cloneDisposable = disposables.add(
                cloneGlobalStyleSheet(style, globalStylesheetClones, targetWindow.document.head)
            );
            disposableStore?.add(cloneDisposable);
        }
        if (shadowRootContainer !== undefined) {
            const cloneDisposable = cloneGlobalStyleSheet(style, globalStylesheetClones, shadowRootContainer);
            disposableStore?.add(cloneDisposable);
        }
    }
    return style;
}
function cloneGlobalStylesheets(targetWindow) {
    const disposables = ( new DisposableStore());
    for (const [globalStylesheet, clonedGlobalStylesheets] of globalStylesheets) {
        disposables.add(
            cloneGlobalStyleSheet(globalStylesheet, clonedGlobalStylesheets, targetWindow.document.head)
        );
    }
    return disposables;
}
function cloneGlobalStyleSheet(globalStylesheet, globalStylesheetClones, windowElement) {
    const disposables = ( new DisposableStore());
    const clone = globalStylesheet.cloneNode(true);
    windowElement.appendChild(clone);
    disposables.add(toDisposable(() => clone.remove()));
    for (const rule of globalStylesheet.sheet?.cssRules ?? []) {
        clone.sheet?.insertRule(rule.cssText, clone.sheet?.cssRules.length);
    }
    disposables.add(sharedMutationObserver.observe(globalStylesheet, disposables, {
        childList: true,
        subtree: isFirefox,
        characterData: isFirefox
    })(() => {
        clone.textContent = globalStylesheet.textContent;
    }));
    globalStylesheetClones.add(clone);
    disposables.add(toDisposable(() => globalStylesheetClones.delete(clone)));
    return disposables;
}
let _sharedStyleSheet = null;
function getSharedStyleSheet() {
    if (!_sharedStyleSheet) {
        _sharedStyleSheet = createStyleSheet();
    }
    return _sharedStyleSheet;
}
function createCSSRule(selector, cssText, style = getSharedStyleSheet()) {
    if (!style || !cssText) {
        return;
    }
    style.sheet?.insertRule(`${selector} {${cssText}}`, 0);
    for (const clonedGlobalStylesheet of globalStylesheets.get(style) ?? []) {
        createCSSRule(selector, cssText, clonedGlobalStylesheet);
    }
}
function removeCSSRulesContainingSelector(ruleName, style = getSharedStyleSheet()) {
    if (!style) {
        return;
    }
    const rules = style.sheet?.cssRules ?? [];
    const toDelete = [];
    for (let i = 0; i < rules.length; i++) {
        const rule = rules[i];
        if (isCSSStyleRule(rule) && rule.selectorText.indexOf(ruleName) !== -1) {
            toDelete.push(i);
        }
    }
    for (let i = toDelete.length - 1; i >= 0; i--) {
        style.sheet?.deleteRule(toDelete[i]);
    }
    for (const clonedGlobalStylesheet of globalStylesheets.get(style) ?? []) {
        removeCSSRulesContainingSelector(ruleName, clonedGlobalStylesheet);
    }
}
function isCSSStyleRule(rule) {
    return typeof rule.selectorText === "string";
}
function createStyleSheetFromObservable(css) {
    const store = ( new DisposableStore());
    const w = store.add(( new WrappedStyleElement()));
    store.add(autorun(reader => {
        w.setStyle(css.read(reader));
    }));
    return store;
}

export { cloneGlobalStylesheets, createCSSRule, createStyleSheet, createStyleSheetFromObservable, isGlobalStylesheet, removeCSSRulesContainingSelector, setContainerElement, shadowRootContainer };
