// @codingame/monaco-vscode-api/vscode/src/vs/workbench/contrib/preferences/browser/settingsLayout.js

import { isWindows, isWeb } from '../../../../base/common/platform.js';
import { localize } from '../../../../nls.js';

const COMMONLY_USED_SETTINGS = [
    "editor.fontSize",
    "editor.formatOnSave",
    "files.autoSave",
    "GitHub.copilot-chat.manageExtension",
    "editor.defaultFormatter",
    "editor.fontFamily",
    "editor.wordWrap",
    "chat.agent.maxRequests",
    "files.exclude",
    "workbench.colorTheme",
    "editor.tabSize",
    "editor.mouseWheelZoom",
    "editor.formatOnPaste"
];
function getCommonlyUsedData(settingGroups) {
    const allSettings = ( new Map());
    for (const group of settingGroups) {
        for (const section of group.sections) {
            for (const s of section.settings) {
                allSettings.set(s.key, s);
            }
        }
    }
    const settings = [];
    for (const id of COMMONLY_USED_SETTINGS) {
        const setting = allSettings.get(id);
        if (setting) {
            settings.push(setting);
        }
    }
    return {
        id: "commonlyUsed",
        label: ( localize(11918, "Commonly Used")),
        settings
    };
}
const tocData = {
    id: "root",
    label: "root",
    children: [{
        id: "editor",
        label: ( localize(11919, "Text Editor")),
        settings: ["editor.*"],
        children: [{
            id: "editor/cursor",
            label: ( localize(11920, "Cursor")),
            settings: ["editor.cursor*"]
        }, {
            id: "editor/find",
            label: ( localize(11921, "Find")),
            settings: ["editor.find.*"]
        }, {
            id: "editor/font",
            label: ( localize(11922, "Font")),
            settings: ["editor.font*"]
        }, {
            id: "editor/format",
            label: ( localize(11923, "Formatting")),
            settings: ["editor.format*"]
        }, {
            id: "editor/diffEditor",
            label: ( localize(11924, "Diff Editor")),
            settings: ["diffEditor.*"]
        }, {
            id: "editor/multiDiffEditor",
            label: ( localize(11925, "Multi-File Diff Editor")),
            settings: ["multiDiffEditor.*"]
        }, {
            id: "editor/minimap",
            label: ( localize(11926, "Minimap")),
            settings: ["editor.minimap.*"]
        }, {
            id: "editor/suggestions",
            label: ( localize(11927, "Suggestions")),
            settings: ["editor.*suggest*"]
        }, {
            id: "editor/files",
            label: ( localize(11928, "Files")),
            settings: ["files.*"]
        }]
    }, {
        id: "workbench",
        label: ( localize(11929, "Workbench")),
        settings: ["workbench.*"],
        children: [{
            id: "workbench/appearance",
            label: ( localize(11930, "Appearance")),
            settings: [
                "workbench.activityBar.*",
                "workbench.*color*",
                "workbench.fontAliasing",
                "workbench.iconTheme",
                "workbench.sidebar.location",
                "workbench.*.visible",
                "workbench.tips.enabled",
                "workbench.tree.*",
                "workbench.view.*"
            ]
        }, {
            id: "workbench/breadcrumbs",
            label: ( localize(11931, "Breadcrumbs")),
            settings: ["breadcrumbs.*"]
        }, {
            id: "workbench/editor",
            label: ( localize(11932, "Editor Management")),
            settings: ["workbench.editor.*"]
        }, {
            id: "workbench/settings",
            label: ( localize(11933, "Settings Editor")),
            settings: ["workbench.settings.*"]
        }, {
            id: "workbench/zenmode",
            label: ( localize(11934, "Zen Mode")),
            settings: ["zenmode.*"]
        }, {
            id: "workbench/screencastmode",
            label: ( localize(11935, "Screencast Mode")),
            settings: ["screencastMode.*"]
        }, {
            id: "workbench/browser",
            label: ( localize(11936, "Browser")),
            settings: ["workbench.browser.*"]
        }]
    }, {
        id: "window",
        label: ( localize(11937, "Window")),
        settings: ["window.*"],
        children: [{
            id: "window/newWindow",
            label: ( localize(11938, "New Window")),
            settings: ["window.*newwindow*"]
        }]
    }, {
        id: "chat",
        label: ( localize(11939, "Chat")),
        children: [{
            id: "chat/agent",
            label: ( localize(11940, "Agent")),
            settings: [
                "chat.agent.*",
                "chat.checkpoints.*",
                "chat.editRequests",
                "chat.requestQueuing.*",
                "chat.undoRequests.*",
                "chat.customAgentInSubagent.*",
                "chat.editing.autoAcceptDelay",
                "chat.editing.confirmEditRequest*",
                "chat.planAgent.defaultModel"
            ]
        }, {
            id: "chat/appearance",
            label: ( localize(11941, "Appearance")),
            settings: [
                "chat.editor.*",
                "chat.fontFamily",
                "chat.fontSize",
                "chat.math.*",
                "chat.agentsControl.*",
                "chat.alternativeToolAction.*",
                "chat.codeBlock.*",
                "chat.editing.explainChanges.enabled",
                "chat.editMode.hidden",
                "chat.editorAssociations",
                "chat.extensionUnification.*",
                "chat.inlineReferences.*",
                "chat.notifyWindow*",
                "chat.statusWidget.*",
                "chat.tips.*",
                "chat.unifiedAgentsBar.*",
                "accessibility.signals.chatUserActionRequired",
                "accessibility.signals.chatResponseReceived"
            ]
        }, {
            id: "chat/sessions",
            label: ( localize(11942, "Sessions")),
            settings: [
                "chat.agentSessionProjection.*",
                "chat.sessions.*",
                "chat.viewProgressBadge.*",
                "chat.viewSessions.*",
                "chat.restoreLastPanelSession",
                "chat.exitAfterDelegation",
                "chat.repoInfo.*"
            ]
        }, {
            id: "chat/tools",
            label: ( localize(11943, "Tools")),
            settings: ["chat.tools.*", "chat.extensionTools.*"]
        }, {
            id: "chat/mcp",
            label: ( localize(11944, "MCP")),
            settings: ["mcp", "chat.mcp.*", "mcp.*"]
        }, {
            id: "chat/context",
            label: ( localize(11945, "Context")),
            settings: [
                "chat.detectParticipant.*",
                "chat.experimental.detectParticipant.*",
                "chat.implicitContext.*",
                "chat.promptFilesLocations",
                "chat.instructionsFilesLocations",
                "chat.modeFilesLocations",
                "chat.agentFilesLocations",
                "chat.agentSkillsLocations",
                "chat.hookFilesLocations",
                "chat.promptFilesRecommendations",
                "chat.useAgentsMdFile",
                "chat.useNestedAgentsMdFiles",
                "chat.useAgentSkills",
                "chat.experimental.useSkillAdherencePrompt",
                "chat.useHooks",
                "chat.includeApplyingInstructions",
                "chat.includeReferencedInstructions",
                "chat.sendElementsToChat.*",
                "chat.useClaudeMdFile"
            ]
        }, {
            id: "chat/inlineChat",
            label: ( localize(11946, "Inline Chat")),
            settings: ["inlineChat.*"]
        }, {
            id: "chat/miscellaneous",
            label: ( localize(11947, "Miscellaneous")),
            settings: ["chat.disableAIFeatures", "chat.allowAnonymousAccess"]
        }]
    }, {
        id: "features",
        label: ( localize(11948, "Features")),
        children: [{
            id: "features/accessibilitySignals",
            label: ( localize(11949, "Accessibility Signals")),
            settings: ["accessibility.signal*"]
        }, {
            id: "features/accessibility",
            label: ( localize(11950, "Accessibility")),
            settings: ["accessibility.*"]
        }, {
            id: "features/explorer",
            label: ( localize(11951, "Explorer")),
            settings: ["explorer.*", "outline.*"]
        }, {
            id: "features/search",
            label: ( localize(11952, "Search")),
            settings: ["search.*"]
        }, {
            id: "features/debug",
            label: ( localize(11953, "Debug")),
            settings: ["debug.*", "launch"]
        }, {
            id: "features/testing",
            label: ( localize(11954, "Testing")),
            settings: ["testing.*"]
        }, {
            id: "features/scm",
            label: ( localize(11955, "Source Control")),
            settings: ["scm.*"]
        }, {
            id: "features/extensions",
            label: ( localize(11956, "Extensions")),
            settings: ["extensions.*"]
        }, {
            id: "features/terminal",
            label: ( localize(11957, "Terminal")),
            settings: ["terminal.*"]
        }, {
            id: "features/task",
            label: ( localize(11958, "Task")),
            settings: ["task.*"]
        }, {
            id: "features/problems",
            label: ( localize(11959, "Problems")),
            settings: ["problems.*"]
        }, {
            id: "features/output",
            label: ( localize(11960, "Output")),
            settings: ["output.*"]
        }, {
            id: "features/comments",
            label: ( localize(11961, "Comments")),
            settings: ["comments.*"]
        }, {
            id: "features/remote",
            label: ( localize(11962, "Remote")),
            settings: ["remote.*"]
        }, {
            id: "features/timeline",
            label: ( localize(11963, "Timeline")),
            settings: ["timeline.*"]
        }, {
            id: "features/notebook",
            label: ( localize(11964, "Notebook")),
            settings: ["notebook.*", "interactiveWindow.*"]
        }, {
            id: "features/mergeEditor",
            label: ( localize(11965, "Merge Editor")),
            settings: ["mergeEditor.*"]
        }, {
            id: "features/issueReporter",
            label: ( localize(11966, "Issue Reporter")),
            settings: ["issueReporter.*"],
            hide: !isWeb
        }]
    }, {
        id: "application",
        label: ( localize(11967, "Application")),
        children: [{
            id: "application/http",
            label: ( localize(11968, "Proxy")),
            settings: ["http.*"]
        }, {
            id: "application/keyboard",
            label: ( localize(11969, "Keyboard")),
            settings: ["keyboard.*"]
        }, {
            id: "application/update",
            label: ( localize(11970, "Update")),
            settings: ["update.*"]
        }, {
            id: "application/telemetry",
            label: ( localize(11971, "Telemetry")),
            settings: ["telemetry.*"]
        }, {
            id: "application/settingsSync",
            label: ( localize(11972, "Settings Sync")),
            settings: ["settingsSync.*"]
        }, {
            id: "application/network",
            label: ( localize(11973, "Network")),
            settings: ["network.*"]
        }, {
            id: "application/experimental",
            label: ( localize(11974, "Experimental")),
            settings: ["application.experimental.*"]
        }, {
            id: "application/other",
            label: ( localize(11975, "Other")),
            settings: ["application.*"],
            hide: isWindows
        }]
    }, {
        id: "security",
        label: ( localize(11976, "Security")),
        settings: ["security.*"],
        children: [{
            id: "security/workspace",
            label: ( localize(11977, "Workspace")),
            settings: ["security.workspace.*"]
        }]
    }]
};

export { getCommonlyUsedData, tocData };
