// @codingame/monaco-vscode-environment-service-override/index.js

import { __decorate, __param } from '@codingame/monaco-vscode-api/external/tslib/tslib.es6';
import { SyncDescriptor } from '@codingame/monaco-vscode-api/vscode/vs/platform/instantiation/common/descriptors';
import { BrowserWorkbenchEnvironmentService } from '@codingame/monaco-vscode-api/vscode/vs/workbench/services/environment/browser/environmentService';
import { IEnvironmentService } from '@codingame/monaco-vscode-api/vscode/vs/platform/environment/common/environment.service';
import '@codingame/monaco-vscode-api/vscode/vs/platform/instantiation/common/instantiation';
import { IProductService } from '@codingame/monaco-vscode-api/vscode/vs/platform/product/common/productService.service';
import { logsPath, getWorkbenchConstructionOptions, getWorkspaceIdentifier } from '@codingame/monaco-vscode-api/workbench';

let InjectedBrowserWorkbenchEnvironmentService = class InjectedBrowserWorkbenchEnvironmentService extends BrowserWorkbenchEnvironmentService {
    constructor(workspaceId = getWorkspaceIdentifier().id, overrides = {}, productService) {
        super(workspaceId, logsPath, getWorkbenchConstructionOptions(), productService);
        this.overrides = overrides;
    }
    get windowLogsPath() {
        return this.overrides.windowLogsPath ?? super.windowLogsPath;
    }
    get logFile() {
        return this.overrides.logFile ?? super.logFile;
    }
    get userRoamingDataHome() {
        return this.overrides.userRoamingDataHome ?? super.userRoamingDataHome;
    }
    get argvResource() {
        return this.overrides.argvResource ?? super.argvResource;
    }
    get cacheHome() {
        return this.overrides.cacheHome ?? super.cacheHome;
    }
    get workspaceStorageHome() {
        return this.overrides.workspaceStorageHome ?? super.workspaceStorageHome;
    }
    get localHistoryHome() {
        return this.overrides.localHistoryHome ?? super.localHistoryHome;
    }
    get stateResource() {
        return this.overrides.stateResource ?? super.stateResource;
    }
};
InjectedBrowserWorkbenchEnvironmentService = __decorate([
    __param(2, IProductService)
], InjectedBrowserWorkbenchEnvironmentService);
function getServiceOverride(overrides) {
    return {
        [IEnvironmentService.toString()]: new SyncDescriptor(InjectedBrowserWorkbenchEnvironmentService, [undefined, overrides], true)
    };
}

export { getServiceOverride as default };
