// vscode/vscode/src/vs/workbench/contrib/terminalContrib/inlineHint/common/terminalInitialHintConfiguration.js

import { localize } from '../../../../../nls.js';
import { TerminalSettingId } from '../../../../../platform/terminal/common/terminal.js';

var TerminalInitialHintSettingId;
(function(TerminalInitialHintSettingId) {
    TerminalInitialHintSettingId["Enabled"] = "terminal.integrated.initialHint";
    TerminalInitialHintSettingId["CopilotCli"] = "terminal.integrated.initialHintCopilotCli";
})(TerminalInitialHintSettingId || (TerminalInitialHintSettingId = {}));
const terminalInitialHintConfiguration = {
    [TerminalInitialHintSettingId.Enabled]: {
        restricted: true,
        markdownDescription: ( localize(
            14282,
            "Controls if the first terminal without input will show a hint about available actions when it is focused. This will only show when {0} is disabled.",
            `\`#${TerminalSettingId.SendKeybindingsToShell}#\``
        )),
        type: "boolean",
        default: true
    },
    [TerminalInitialHintSettingId.CopilotCli]: {
        restricted: true,
        markdownDescription: ( localize(
            14283,
            "When enabled, the terminal initial hint will suggest using Copilot CLI by typing {0} instead of opening Copilot Chat.",
            "`copilot`"
        )),
        type: "boolean",
        default: false,
        tags: ["experimental"],
        experiment: {
            mode: "auto"
        }
    }
};

export { TerminalInitialHintSettingId, terminalInitialHintConfiguration };
