// @codingame/monaco-vscode-api/vscode/src/vs/workbench/contrib/welcomeOnboarding/common/onboardingTypes.js

import { localize } from '../../../../nls.js';
import { isMacintosh } from '../../../../base/common/platform.js';

var OnboardingStepId;
(function(OnboardingStepId) {
    OnboardingStepId["SignIn"] = "onboarding.signIn";
    OnboardingStepId["Personalize"] = "onboarding.personalize";
    OnboardingStepId["AiPreference"] = "onboarding.aiPreference";
    OnboardingStepId["AgentSessions"] = "onboarding.agentSessions";
})(OnboardingStepId || (OnboardingStepId = {}));
function getOnboardingStepTitle(stepId) {
    switch (stepId) {
    case OnboardingStepId.SignIn:
        return localize(15700, "Sign In");
    case OnboardingStepId.Personalize:
        return localize(15701, "Make It Yours");
    case OnboardingStepId.AiPreference:
        return localize(15702, "Your AI Style");
    case OnboardingStepId.AgentSessions:
        return localize(15703, "Meet Your Agentic Coding Partner");
    }
}
function getOnboardingStepSubtitle(stepId) {
    switch (stepId) {
    case OnboardingStepId.SignIn:
        return localize(15704, "Sync settings, unlock AI features, and connect to GitHub");
    case OnboardingStepId.Personalize:
        return localize(15705, "Choose your theme and keyboard mapping");
    case OnboardingStepId.AiPreference:
        return localize(15706, "Choose how much AI collaboration fits your workflow");
    case OnboardingStepId.AgentSessions:
        return localize(15707, "Tip: Press {0} to open Chat", isMacintosh ? "⌘⌃I" : "Ctrl+Alt+I");
    }
}
const ONBOARDING_STEPS = [
    OnboardingStepId.SignIn,
    OnboardingStepId.Personalize,
    OnboardingStepId.AgentSessions
];
var AiCollaborationMode;
(function(AiCollaborationMode) {
    AiCollaborationMode["CodeFirst"] = "code-first";
    AiCollaborationMode["Balanced"] = "balanced";
    AiCollaborationMode["AgentForward"] = "agent-forward";
})(AiCollaborationMode || (AiCollaborationMode = {}));
const ONBOARDING_AI_PREFERENCE_OPTIONS = [{
    id: AiCollaborationMode.CodeFirst,
    label: ( localize(15708, "I Write the Code")),
    description: ( localize(
        15709,
        "AI assists with suggestions and answers questions when you ask. You stay in control of every edit."
    )),
    icon: "edit"
}, {
    id: AiCollaborationMode.Balanced,
    label: ( localize(15710, "Side by Side")),
    description: ( localize(
        15711,
        "Inline suggestions plus a chat panel for deeper collaboration. A balance of writing and delegating."
    )),
    icon: "layoutSidebarRight"
}, {
    id: AiCollaborationMode.AgentForward,
    label: ( localize(15712, "AI Takes the Lead")),
    description: ( localize(
        15713,
        "Let the agent drive — describe what you want and review the result. Great for scaffolding and exploration."
    )),
    icon: "copilot"
}];
const ONBOARDING_STORAGE_KEY = "welcomeOnboarding.state";

export { AiCollaborationMode, ONBOARDING_AI_PREFERENCE_OPTIONS, ONBOARDING_STEPS, ONBOARDING_STORAGE_KEY, OnboardingStepId, getOnboardingStepSubtitle, getOnboardingStepTitle };
