// vscode/vscode/src/vs/platform/actionWidget/browser/actionWidget.js

import { registerCss } from '../../../../../../css.js';
import { __decorate, __param } from '../../../../../../external/tslib/tslib.es6.js';
import { createElement, addDisposableGenericMouseDownListener, addDisposableListener, EventType, trackFocus, getActiveElement, $ } from '../../../base/browser/dom.js';
import { ActionBar } from '../../../base/browser/ui/actionbar/actionbar.js';
import { KeyMod as KeyMod$1, KeyCode } from '../../../base/common/keyCodes.js';
import { Disposable, MutableDisposable, DisposableStore } from '../../../base/common/lifecycle.js';
import * as actionWidget from './actionWidget.css';
import { localize, localize2 } from '../../../nls.js';
import { ActionList, acceptSelectedActionCommand, previewSelectedActionCommand } from './actionList.js';
import { registerAction2, Action2 } from '../../actions/common/actions.js';
import { RawContextKey, ContextKeyExpr } from '../../contextkey/common/contextkey.js';
import { IContextKeyService } from '../../contextkey/common/contextkey.service.js';
import { IContextViewService } from '../../contextview/browser/contextView.service.js';
import '../../instantiation/common/extensions.js';
import { IInstantiationService } from '../../instantiation/common/instantiation.js';
import { KeybindingWeight } from '../../keybinding/common/keybindingsRegistry.js';
import { registerColor } from '../../theme/common/colorUtils.js';
import '../../theme/common/colors/baseColors.js';
import '../../theme/common/colors/chartsColors.js';
import '../../theme/common/colors/editorColors.js';
import { inputActiveOptionBackground } from '../../theme/common/colors/inputColors.js';
import '../../theme/common/colors/listColors.js';
import '../../theme/common/colors/menuColors.js';
import '../../theme/common/colors/minimapColors.js';
import '../../theme/common/colors/miscColors.js';
import '../../theme/common/colors/quickpickColors.js';
import '../../theme/common/colors/searchColors.js';
import { IActionWidgetService } from './actionWidget.service.js';

registerCss(actionWidget);
registerColor("actionBar.toggledBackground", inputActiveOptionBackground, ( localize(1786, "Background color for toggled action items in action bar.")));
const ActionWidgetContextKeys = {
    Visible: ( new RawContextKey("codeActionMenuVisible", false, ( localize(1787, "Whether the action widget list is visible")))),
    FilterFocused: ( new RawContextKey("codeActionMenuFilterFocused", false, ( localize(1788, "Whether the action widget filter input is focused"))))
};
let ActionWidgetService = class ActionWidgetService extends Disposable {
    get isVisible() {
        return ActionWidgetContextKeys.Visible.getValue(this._contextKeyService) || false;
    }
    constructor(_contextViewService, _contextKeyService, _instantiationService) {
        super();
        this._contextViewService = _contextViewService;
        this._contextKeyService = _contextKeyService;
        this._instantiationService = _instantiationService;
        this._list = this._register(( new MutableDisposable()));
    }
    show(
        user,
        supportsPreview,
        items,
        delegate,
        anchor,
        container,
        actionBarActions,
        accessibilityProvider,
        listOptions
    ) {
        const visibleContext = ActionWidgetContextKeys.Visible.bindTo(this._contextKeyService);
        const list = this._instantiationService.createInstance(
            ActionList,
            user,
            supportsPreview,
            items,
            delegate,
            accessibilityProvider,
            listOptions,
            anchor
        );
        this._contextViewService.showContextView({
            getAnchor: () => anchor,
            render: container => {
                visibleContext.set(true);
                return this._renderWidget(container, list, actionBarActions ?? []);
            },
            onHide: didCancel => {
                visibleContext.reset();
                this._onWidgetClosed(didCancel);
            },
            get anchorPosition() {
                return list.anchorPosition;
            }
        }, container, false);
    }
    acceptSelected(preview) {
        this._list.value?.acceptSelected(preview);
    }
    focusPrevious() {
        this._list?.value?.focusPrevious();
    }
    focusNext() {
        this._list?.value?.focusNext();
    }
    collapseSection() {
        this._list?.value?.collapseFocusedSection();
    }
    expandSection() {
        this._list?.value?.expandFocusedSection();
    }
    toggleSection() {
        return this._list?.value?.toggleFocusedSection() ?? false;
    }
    clearFilter() {
        return this._list?.value?.clearFilter() ?? false;
    }
    hide(didCancel) {
        this._list.value?.hide(didCancel);
        this._list.clear();
    }
    clear() {
        this._list.clear();
    }
    _renderWidget(element, list, actionBarActions) {
        const widget = createElement("div");
        widget.classList.add("action-widget");
        element.appendChild(widget);
        this._list.value = list;
        if (this._list.value) {
            if (this._list.value.filterContainer) {
                widget.appendChild(this._list.value.filterContainer);
            }
            widget.appendChild(this._list.value.domNode);
        } else {
            throw ( new Error("List has no value"));
        }
        const renderDisposables = ( new DisposableStore());
        const menuBlock = createElement("div");
        const block = element.appendChild(menuBlock);
        block.classList.add("context-view-block");
        renderDisposables.add(addDisposableGenericMouseDownListener(block, e => e.stopPropagation()));
        const pointerBlockDiv = createElement("div");
        const pointerBlock = element.appendChild(pointerBlockDiv);
        pointerBlock.classList.add("context-view-pointerBlock");
        renderDisposables.add(
            addDisposableListener(pointerBlock, EventType.POINTER_MOVE, () => pointerBlock.remove())
        );
        renderDisposables.add(
            addDisposableGenericMouseDownListener(pointerBlock, () => pointerBlock.remove())
        );
        let actionBarWidth = 0;
        if (actionBarActions.length) {
            const actionBar = this._createActionBar(".action-widget-action-bar", actionBarActions);
            if (actionBar) {
                widget.appendChild(actionBar.getContainer().parentElement);
                renderDisposables.add(actionBar);
                actionBarWidth = actionBar.getContainer().offsetWidth;
            }
        }
        const width = this._list.value?.layout(actionBarWidth);
        widget.style.width = `${width}px`;
        this._list.value?.focus();
        const filterFocusedContext = ActionWidgetContextKeys.FilterFocused.bindTo(this._contextKeyService);
        renderDisposables.add({
            dispose: () => filterFocusedContext.reset()
        });
        if (this._list.value?.filterInput) {
            const filterInput = this._list.value.filterInput;
            renderDisposables.add(
                addDisposableListener(filterInput, "focus", () => filterFocusedContext.set(true))
            );
            renderDisposables.add(
                addDisposableListener(filterInput, "blur", () => filterFocusedContext.set(false))
            );
        }
        const focusTracker = renderDisposables.add(trackFocus(element));
        renderDisposables.add(focusTracker.onDidBlur(() => {
            const activeElement = getActiveElement();
            if (activeElement?.closest(".action-widget-hover") || activeElement?.closest(".action-list-submenu-panel")) {
                return;
            }
            this.hide(true);
        }));
        return renderDisposables;
    }
    _createActionBar(className, actions) {
        if (!actions.length) {
            return undefined;
        }
        const container = $(className);
        const actionBar = ( new ActionBar(container));
        actionBar.push(actions, {
            icon: false,
            label: true
        });
        return actionBar;
    }
    _onWidgetClosed(didCancel) {
        this._list.value?.hide(didCancel);
    }
};
ActionWidgetService = ( __decorate([( __param(0, IContextViewService)), ( __param(1, IContextKeyService)), ( __param(2, IInstantiationService))], ActionWidgetService));
const weight = KeybindingWeight.EditorContrib + 1000;
registerAction2(class extends Action2 {
    constructor() {
        super({
            id: "hideCodeActionWidget",
            title: ( localize2(1789, "Hide action widget")),
            precondition: ActionWidgetContextKeys.Visible,
            keybinding: {
                weight,
                primary: KeyCode.Escape,
                secondary: [KeyMod$1.Shift | KeyCode.Escape]
            }
        });
    }
    run(accessor) {
        accessor.get(IActionWidgetService).hide(true);
    }
});
registerAction2(class extends Action2 {
    constructor() {
        super({
            id: "clearFilterCodeActionWidget",
            title: ( localize2(1790, "Clear action widget filter")),
            precondition: ( ContextKeyExpr.and(ActionWidgetContextKeys.Visible, ActionWidgetContextKeys.FilterFocused)),
            keybinding: {
                weight: weight + 1,
                primary: KeyCode.Escape
            }
        });
    }
    run(accessor) {
        const widgetService = accessor.get(IActionWidgetService);
        if (widgetService instanceof ActionWidgetService) {
            if (!widgetService.clearFilter()) {
                widgetService.hide(true);
            }
        }
    }
});
registerAction2(class extends Action2 {
    constructor() {
        super({
            id: "selectPrevCodeAction",
            title: ( localize2(1791, "Select previous action")),
            precondition: ActionWidgetContextKeys.Visible,
            keybinding: {
                weight,
                primary: KeyCode.UpArrow,
                secondary: [KeyMod$1.CtrlCmd | KeyCode.UpArrow],
                mac: {
                    primary: KeyCode.UpArrow,
                    secondary: [KeyMod$1.CtrlCmd | KeyCode.UpArrow, KeyMod$1.WinCtrl | KeyCode.KeyP]
                }
            }
        });
    }
    run(accessor) {
        const widgetService = accessor.get(IActionWidgetService);
        if (widgetService instanceof ActionWidgetService) {
            widgetService.focusPrevious();
        }
    }
});
registerAction2(class extends Action2 {
    constructor() {
        super({
            id: "selectNextCodeAction",
            title: ( localize2(1792, "Select next action")),
            precondition: ActionWidgetContextKeys.Visible,
            keybinding: {
                weight,
                primary: KeyCode.DownArrow,
                secondary: [KeyMod$1.CtrlCmd | KeyCode.DownArrow],
                mac: {
                    primary: KeyCode.DownArrow,
                    secondary: [KeyMod$1.CtrlCmd | KeyCode.DownArrow, KeyMod$1.WinCtrl | KeyCode.KeyN]
                }
            }
        });
    }
    run(accessor) {
        const widgetService = accessor.get(IActionWidgetService);
        if (widgetService instanceof ActionWidgetService) {
            widgetService.focusNext();
        }
    }
});
registerAction2(class extends Action2 {
    constructor() {
        super({
            id: "collapseSectionCodeAction",
            title: ( localize2(1793, "Collapse section")),
            precondition: ( ContextKeyExpr.and(ActionWidgetContextKeys.Visible, ( ActionWidgetContextKeys.FilterFocused.negate()))),
            keybinding: {
                weight,
                primary: KeyCode.LeftArrow
            }
        });
    }
    run(accessor) {
        const widgetService = accessor.get(IActionWidgetService);
        if (widgetService instanceof ActionWidgetService) {
            widgetService.collapseSection();
        }
    }
});
registerAction2(class extends Action2 {
    constructor() {
        super({
            id: "expandSectionCodeAction",
            title: ( localize2(1794, "Expand section")),
            precondition: ( ContextKeyExpr.and(ActionWidgetContextKeys.Visible, ( ActionWidgetContextKeys.FilterFocused.negate()))),
            keybinding: {
                weight,
                primary: KeyCode.RightArrow
            }
        });
    }
    run(accessor) {
        const widgetService = accessor.get(IActionWidgetService);
        if (widgetService instanceof ActionWidgetService) {
            widgetService.expandSection();
        }
    }
});
registerAction2(class extends Action2 {
    constructor() {
        super({
            id: "toggleSectionCodeAction",
            title: ( localize2(1795, "Toggle section")),
            precondition: ( ContextKeyExpr.and(ActionWidgetContextKeys.Visible, ( ActionWidgetContextKeys.FilterFocused.negate()))),
            keybinding: {
                weight,
                primary: KeyCode.Space
            }
        });
    }
    run(accessor) {
        const widgetService = accessor.get(IActionWidgetService);
        if (widgetService instanceof ActionWidgetService) {
            if (!widgetService.toggleSection()) {
                widgetService.acceptSelected();
            }
        }
    }
});
registerAction2(class extends Action2 {
    constructor() {
        super({
            id: acceptSelectedActionCommand,
            title: ( localize2(1796, "Accept selected action")),
            precondition: ActionWidgetContextKeys.Visible,
            keybinding: {
                weight,
                primary: KeyCode.Enter,
                secondary: [KeyMod$1.CtrlCmd | KeyCode.Period]
            }
        });
    }
    run(accessor) {
        const widgetService = accessor.get(IActionWidgetService);
        if (widgetService instanceof ActionWidgetService) {
            widgetService.acceptSelected();
        }
    }
});
registerAction2(class extends Action2 {
    constructor() {
        super({
            id: previewSelectedActionCommand,
            title: ( localize2(1797, "Preview selected action")),
            precondition: ActionWidgetContextKeys.Visible,
            keybinding: {
                weight,
                primary: KeyMod$1.CtrlCmd | KeyCode.Enter
            }
        });
    }
    run(accessor) {
        const widgetService = accessor.get(IActionWidgetService);
        if (widgetService instanceof ActionWidgetService) {
            widgetService.acceptSelected(true);
        }
    }
});

export { ActionWidgetService };
