// @codingame/monaco-vscode-api/vscode/src/vs/base/browser/overlayLayoutElement.js

import { createElement, setParentFlowTo } from './dom.js';
import { Lazy } from '../common/lazy.js';
import { generateUuid } from '../common/uuid.js';

const supportsAnchorPositioning = ( new Lazy(() => CSS.supports("(top: anchor(top))")));
function getOrCreateAnchorName(element) {
    const existing = element.style.getPropertyValue("anchor-name");
    if (existing) {
        return existing;
    }
    const name = `--overlay-anchor-${generateUuid()}`;
    element.style.setProperty("anchor-name", name);
    return name;
}
class OverlayLayoutElement {
    constructor() {
        this.content = createElement("div");
        this.content.style.position = "absolute";
        this.content.style.overflow = "hidden";
        this._root = createElement("div");
        this._root.appendChild(this.content);
        this.reapplyLayoutStyles();
    }
    reapplyLayoutStyles() {
        if (supportsAnchorPositioning.value) {
            this.content.style.position = "fixed";
            this.content.style.top = "anchor(top)";
            this.content.style.left = "anchor(left)";
            this.content.style.width = "anchor-size(width)";
            this.content.style.height = "anchor-size(height)";
            this.content.style.pointerEvents = "auto";
        }
        this._root.style.position = "absolute";
        this._root.style.pointerEvents = "none";
    }
    dispose() {
        this.root.remove();
    }
    get root() {
        return this._root;
    }
    layoutOverAnchorElement(anchorElement, options) {
        if (supportsAnchorPositioning.value) {
            if (this._currentAnchor?.element !== anchorElement) {
                const name = getOrCreateAnchorName(anchorElement);
                this.content.style.setProperty("position-anchor", name);
                setParentFlowTo(this.content, anchorElement);
                this._currentAnchor = {
                    element: anchorElement,
                    name
                };
            }
        } else {
            const cs = this.content.style;
            if (options?.fallbackPosition) {
                cs.top = `${options.fallbackPosition.top}px`;
                cs.left = `${options.fallbackPosition.left}px`;
            } else {
                const anchorRect = anchorElement.getBoundingClientRect();
                const parentRect = this.content.parentElement.getBoundingClientRect();
                const parentBorderTop = (parentRect.height - this.content.parentElement.clientHeight) / 2.0;
                const parentBorderLeft = (parentRect.width - this.content.parentElement.clientWidth) / 2.0;
                cs.top = `${anchorRect.top - parentRect.top - parentBorderTop}px`;
                cs.left = `${anchorRect.left - parentRect.left - parentBorderLeft}px`;
            }
            setParentFlowTo(this.content, anchorElement);
            const anchorRect = anchorElement.getBoundingClientRect();
            cs.width = `${options?.fallbackDimension ? options.fallbackDimension.width : anchorRect.width}px`;
            cs.height = `${options?.fallbackDimension ? options.fallbackDimension.height : anchorRect.height}px`;
        }
        this._updateClipping(anchorElement, options?.clippingContainer);
    }
    _updateClipping(anchorElement, clippingContainer) {
        if (supportsAnchorPositioning.value) {
            if (this._clippingAnchor?.element === clippingContainer) {
                return;
            }
            this._root.style.removeProperty("position-anchor");
            const ws = this._root.style;
            if (clippingContainer) {
                const name = getOrCreateAnchorName(clippingContainer);
                ws.clipPath = "content-box";
                ws.setProperty("position-anchor", name);
                ws.setProperty("top", "anchor(top)");
                ws.setProperty("left", "anchor(left)");
                ws.setProperty("width", `anchor-size(width)`);
                ws.setProperty("height", `anchor-size(height)`);
                this._clippingAnchor = {
                    element: clippingContainer,
                    name
                };
            } else {
                ws.clipPath = "";
                ws.setProperty("top", "0");
                ws.setProperty("left", "0");
                ws.setProperty("right", "0");
                ws.setProperty("bottom", "0");
                this._clippingAnchor = undefined;
            }
        } else {
            if (clippingContainer) {
                const anchorRect = anchorElement.getBoundingClientRect();
                const clipRect = clippingContainer.getBoundingClientRect();
                const top = Math.max(clipRect.top - anchorRect.top, 0);
                const right = Math.max(anchorRect.width - (anchorRect.right - clipRect.right), 0);
                const bottom = Math.max(anchorRect.height - (anchorRect.bottom - clipRect.bottom), 0);
                const left = Math.max(clipRect.left - anchorRect.left, 0);
                this.content.style.clipPath = `polygon(${left}px ${top}px, ${right}px ${top}px, ${right}px ${bottom}px, ${left}px ${bottom}px)`;
                this._clippingAnchor = {
                    element: clippingContainer,
                    name: ""
                };
            } else {
                this.content.style.clipPath = "";
                this._clippingAnchor = undefined;
            }
        }
    }
}

export { OverlayLayoutElement };
